### ISSUE TITLE
Add support for using raw (unescaped) HTML

### ISSUE THREAD

**Issue by doekman** [stefanhaustein/notemplate#6]
If you use a string as a node, it is interpreted as plain text, and all special HTML codes are escaped.

Sometimes, it is useful to insert a string, as if it were HTML. Example: if you want to use a Markdown library with _notemplate_, this   would be nice.

Example:

    tag.p(tag.RAW("This <code>string</code> contains text &amp; HTML."));

Not sure if `RAW` is the name to go with, but it is short. Others are `HTML`or `INNER_HTML`. The capitals are an indication that it's not an ordinary tag-function.

Another idea is to use [tagged template literals](dfile:///Users/doekman/Library/Application%20Support/Dash/DocSets/JavaScript/JavaScript.docset/Contents/Resources/Documents/developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals.html#tagged_templates):

    tag.p(tag`This <code>string</code> contains text &amp; HTML.`));

This looks like a perfect fit. Not 100% sure, but I think variables should NOT automatically be escaped:

    let contents = "text &amp; HTML";
    tag.p(tag`This <code>string</code> contains HTML: <code>${contents}</code>.`));

To the previous line would produce the same DOM as the following line:

    tag.p("This ",
        tag.code("string"),
        " contains HTML: ",
        tag.code("text &amp; HTML"))

I think both interfaces can come into use. The `tag.RAW()` function works with variables, and the `tag\`\`` string just types/reads more natural.

If you like this, I can make a PR.

**Comment by stefanhaustein** [comment]
Hey sorry for the delay... I think the main problem is that it's tricky to "mix in" with other content.... How would the function work if there is other content, too?

Maybe an alternative would be a function that sets innerHtml and returns the node, so it could still occur when building a tree using functions?

Something like
```
innerHtml(tag.p({"class": "foo"}), "This <code>string</code> contains text &amp; HTML.")
```

It should be rare after all?

An alternative might be to make more explicit that it overwrites all content (so far)?

Or we could define it as function on the element somehow, e.g.

```
tag.p.withInnerHtml("This <code>string</code> contains text &amp; HTML.")({"class": "foo"})
```
 
That said, shouldn't innerHtml just be avoided as far as possible? Not sure we should enocurage it...

**Comment by doekman** [comment]
Hi, no sweat about the delay. This is "just" an open source project. For me it is a side project for work, so I also might not respond very quickly.
 
I came up with this functionality because I needed it. I've made a start with an implementation (it's not ready yet), see here:

https://github.com/doekman/notemplate/blob/raw-html/src/notemplate.js

About your thoughts of innerHtml te be avoided: my thoughts are, it exists, so why not let the user of the library decide whether it's a good idea or not?

About the implementation: I'm curious if it is much slower compared compared to proper node creation. InnerHtml used to be (IE6 times) quicker, especially when creating complex trees. 

Also: I'm using `element.innerHTML += ...`. If this extension to the library is being considered, I might also try another way to do this, and see if I can benchmark this. The innerHtml property is probably heavily optimised by browser vendors...

About the interface: I'm going to think about it. I do like your `p.withInnerHtml` idea.

**Comment by stefanhaustein** [comment]
Hey sorry I just didn't realize you can just append to innerHtml ... O:) 

The tagged template idea looks interesting, but I think it would be more useful if it would escape -- checking for contained DOM nodes, which would be appended unescaped (depending on implementation details via outerHtml)?

Maybe we can find a better name for it? Using tag seems odd as the template doesn't necessarily create a tag / element?

How about raw? We'd have two top level functions instead of one but maybe it would be a nice complement :)

p.s. How about letting "raw" produce a DocumentFragment? This way we could still use raw inside raw templates -- and probably also support raw() with a regular non-template string parameter for including externally produced raw strings?

**Comment by doekman** [comment]
> The tagged template idea looks interesting, but I think it would be more useful if it would escape -- checking for contained DOM nodes, which would be appended unescaped (depending on implementation details via outerHtml)?

Not sure if I understand correctly. Something like this?

```javascript:
let code_string = tag.code("string");
let contents = "text &amp; HTML";
tag.p(raw`This ${code_string} contains HTML: <code>${contents}</code>.`));
```
Results in:

```html
<p>
    This <code>string</code> contains HTML: <code>text &amp;amp; HTML</code>.
</p>
```

Adding non-string types in raw-strings will be interesting; however, I can't think of any use-cases at the moment. 

Yeah, `raw` sounds nice. Did you remove the yin/yang from the issue (it's still there in my email notification)? I liked it.

DocumentFragment: I knew it existed... Seems like a perfect fit.

**Comment by stefanhaustein** [comment]
Yes -- and a minimal change can fix the double-escaping:

```
let code_string = tag.code("string");
let contents = raw`text &amp; HTML`;
tag.p(raw`This ${code_string} contains HTML: <code>${contents}</code>.`));
```
Results in:
```
<p>
    This <code>string</code> contains HTML: <code>text &amp; HTML</code>.
</p>
```

The reason why I think we should escape by default is that there seem to be potential use cases with escaping and it's easy to prevent it using raw. 

tag.p(raw`Name: ${name}<br />Address: ${address}<br />Town: ${town}`)

It just seems cleaner to "error" on the safe side....

I had removed the yin/yang because I wasn't sure I like it and the attachment was quite big/distracting. Having some sort of logo is probably a separate discussion...

**Comment by doekman** [comment]
> Yes -- and a minimal change can fix the double-escaping:
> 
> ```
> let code_string = tag.code("string");
> let contents = raw`text &amp; HTML`;
> tag.p(raw`This ${code_string} contains HTML: <code>${contents}</code>.`));
> ```
> 
> Results in:
> 
> ```
> <p>
>     This <code>string</code> contains HTML: <code>text &amp; HTML</code>.
> </p>
> ```
> 
> The reason why I think we should escape by default is that there seem to be potential use cases with escaping and it's easy to prevent it using raw.
> 
> tag.p(raw`Name: ${name}<br />Address: ${address}<br />Town: ${town}`)
> 
> It just seems cleaner to "error" on the safe side....
> 
> I had removed the yin/yang because I wasn't sure I like it and the attachment was quite big/distracting. Having some sort of logo is probably a separate discussion...

**Comment by doekman** [comment]
Oops, pressed the wrong button 🙈.

About wording: with escaping I mean setting `innerText` while you probably mean `innerHTML`. That is what I meant with `&amp;amp;` (the `&` sign will be escaped to `&amp;` and the `amp;` part is appended as text).

But to be safe: we interpret string-variables in raw-strings as is was raw HTML. 

I will make changes to my feature branch shortly, not sure if it will be this week.

**Comment by stefanhaustein** [comment]
Note that I changed the variable value to raw(x) to
suppress double-escaping in my example

Both examples (yours and mine) should be consistent.

Inserted strings should be escaped, as this seems to be a use case for
templates where names / text fields are inserted

Escaping can be avoided by using the raw() function, turning the content
into a DOM fragment, which will be inserted directly.





Am Do., 19. Mai 2022 um 09:55 Uhr schrieb Doeke Zanstra <
***@***.***>:

**Comment by stefanhaustein** [comment]
By escaping I mean "innerText" -- innerHtml inserts "raw" content. 

Note that I had changed line 2 of the example (let contents = ...), adding a raw call, demonstrating how it's still possible to insert "raw" unescaped content (via fragment generation) using placeholders in a "raw" template.

**Comment by doekman** [comment]
I noticed I get a bit confused by the term _escaping_, and find it hard to follow. I guess this is because it is an imperative term: it describes how things are done. I think it is more clear if we talk about text- and html-strings. In html-strings we can expect control codes like `<code>` (tag open) and entities `&lt;` (the `<` symbol).

In code html-strings are assigned to innerHTML and text-strings to innerText. To apply content "types" to strings always helped me reason about it much better (like embedded SQL).

Let me add string "types" in our examples:

```javascript:
let code_string = tag.code("string");
//                          ^^^^^^ text-string
let contents = raw`text &amp; HTML`;
//                 ^^^^^^^^^^^^^^^ html-string
tag.p(raw`This ${code_string} contains HTML: <code>${contents}</code>.`));
//        ^^^^^              ^^^^^^^^^^^^^^^^^^^^^^           ^^^^^^^^ html-strings
//               ^^^^^^^^^^^                         ^^^^^^^^ variables of type Node

let name = "Doeke <strong>Zanstra</strong>";
//          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ html-string
let address = "JavaScript Drive 12";
//             ^^^^^^^^^^^^^^^^^^^ text-string
let town = "Scripting Town";
//          ^^^^^^^^^^^^^^ text-string
tag.p(raw`Name: ${raw(name)}<br />Address: ${address}<br />Town: ${town}`)
//        ^^^^^^            ^^^^^^^^^^^^^^^          ^^^^^^^^^^^^           html-strings
//                    ^^^^ variable with string needs to be explicitly marked as html
//                ^^^^^^^^^ function result is of type Node
//                                           ^^^^^^^^              ^^^^     strings handled as text-strings
```

So basically all strings are considered text-strings (or plain/text). If you want to use html-strings (or text/html), the string must directly be specified in a raw-string or a raw-funcion call.

Notice I altered your example (last line above) to demonstrate how to use the raw-function.

Does this make sense? And is this what you mean?

**Comment by stefanhaustein** [comment]
Yes, exactly!

**Comment by doekman** [comment]
So I fixed the raw-handling, and added an example.

Next is enhance some of the readme.

A thought: the `raw`-function could be published as an external module. This way, we would better fit into the npm-ecology... 😉
